Port ngrx/entity and add createAsyncThunk (#352)
reduxjs/redux-toolkit: docs/api/createAsyncThunk.md

      state.entities.push(action.payload)
    }
  }
})

// Later, dispatch the thunk as needed in the app
dispatch(fetchUserById(123))
```

## Parameters

### `type`

A string that will be used to generate additional Redux action type constants, representing the lifecycle of an async request:

For example, a `type` argument of `'users/requestStatus'` will generate these action types:

- `pending`: `'users/requestStatus/pending'`
- `fulfilled`: `'users/requestStatus/fulfilled'`
- `rejected`: `'users/requestStatus/rejected'`

### `payloadCreator`

A callback function that should return a promise containing the result of some asynchronous logic. It may also return a value synchronously. If there is an error, it should return a rejected promise containing either an `Error` instance or a plain value such as a descriptive error message.

The `payloadCreator` function can contain whatever logic you need to calculate an appropriate result. This could include a standard AJAX data fetch request, multiple AJAX calls with the results combined into a final value, interactions with React Native `AsyncStorage`, and so on.

The `payloadCreator` function will be called with two arguments:

> - `arg`: a single value, containing the first parameter that was passed to the thunk action creator when it was dispatched. This is useful for passing in values like item IDs that may be needed as part of the request. If you need to pass in multiple values, pass them together in an object when you dispatch the thunk, like `dispatch(fetchUsers({status: 'active', sortBy: 'name'}))`.

Reviewer: Suggestion: treat the object as the default approach. Otherwise it is tempting to add positional args after the first one, which will not work, but you might not even notice for a while depending on how defaults are handled.

```suggestion
`arg`: a single value, containing the first parameter that was passed to the thunk action creator when it was dispatched. This is useful for passing in values like item IDs that may be needed as part of the request. Passing an object is recommended for most cases,
```

Author: Are you talking about at strictly a documentation level, or something at the code level too?

Reviewer: Just docs

  - `extra`: the "extra argument" given to the thunk middleware on setup, if available
  - `requestId`: a unique string ID value that was automatically generated to identify this request sequence
  - `signal`: an [`AbortController.signal` object](https://developer.mozilla.org/en-US/docs/Web/API/AbortController/signal) that may be used to see if another part of the app logic has marked this request as needing cancelation.

The logic in the `payloadCreator` function may use any of these values as needed to calculate the result.

## Return Value

`createAsyncThunk` returns a standard Redux thunk action creator. The thunk action creator function will have plain action creators for the `pending`, `fulfilled`, and `rejected` cases attached as nested fields.

When dispatched, the thunk will:

- dispatch the `pending` action
- call the `payloadCreator` callback and wait for the returned promise to settle
- when the promise settles:
  - if the promise resolved successfully, dispatch the `fulfilled` action with the promise value as `action.payload`
  - if the promise failed, dispatch the `rejected` action with a serialized version of the error value as `action.error`

> - Return a fulfilled promise containing the final dispatched action (either the `fulfilled` or `rejected` action object)

Reviewer: API comment: if awaiting the thunk returns a usable payload, it's very tempting to use the value directly (for example, by passing it to a local setState) instead of reading from redux state. I've seen this is a lot in an app that has a similar utility.

Author: You can already do that with thunks today, so I don't see any difference here conceptually.

Most of the time, folks are just interested in kicking off some kind of additional behavior after the thunk promise resolves:

```js
dispatch(someThunk())
  .then(() => {})
```

The issue here is that since we're catching errors internally, we always return a fulfilled promise, and it's hypothetically possible the user _might_ A) be interested in the original API call result, or B) want to deal with the API call having rejected.

So, we're trying to provide a way to enable that.
